corporate/lib/stripe.py

```diff
@@ -4440,7 +4447,9 @@ def get_plan_renewal_or_end_date(plan: CustomerPlan, event_time: datetime) -> da
 def invoice_plans_as_needed(event_time: Optional[datetime] = None) -> None:
     if event_time is None:  # nocoverage
         event_time = timezone_now()
-    for plan in CustomerPlan.objects.filter(next_invoice_date__lte=event_time):
+    # For self hosted legacy plan with status SWITCH_PLAN_TIER_AT_PLAN_END, we need
+    # to invoice legacy plan followed by new plan on the same day, hence ordered by ID.
+    for plan in CustomerPlan.objects.filter(next_invoice_date__lte=event_time).order_by("id"):
         remote_server: Optional[RemoteZulipServer] = None
```

Comment:
This works, I wonder if we can do something better here. Like first process all the non-ACTIVE plans and then process all the ACTIVE plans.

Reply:
Hmm. This logic of sorting by plan ID does seem fairly stable/correct, I think? New plans will always have a newer ID, and we're never going to reactivate an ended plan instead of just making a new one with the same tier.

The main things I'd worry about are:
* This loop does the query once, so I think if `make_end_of_cycle_updates_as_needed` causes there to be additional plans that match this query, we need something sane to happen.

Reply:
While processing a plan, if we update `next_invoice_date` or `end_date` of another plan to be `<= event_time` -- so this is a similar case to legacy plan scheduled to upgrade. In such cases our logic should be to set those values outside of `make_end_of_cycle...` if we already know the date & order in which the plans should be invoiced?

But yeah, there could be some situations where those values can't be pre-known & we need to set it inside `make_end_of_cycle...`; in such cases should we enqueue this job to run again immediately? -- I think the perfect answer to this will depend when we actually encounter one such case?

Reply:
Yeah I think there's probably nothing needed for this PR on that point.